### DISCUSSION TITLE
Erweiterung/Frage: Import von Artikeldaten/Preisen

### DISCUSSION THREAD

**Discussion by schattenlieger** [dragotin/kraft#172] [Ideas]
Guten Tag,
Ich habe Kraft vor einigen Tagen entdeckt, um das manuelle Schreiben unter Softmaker Office etwas zu automatieren. 
Ausgewachsene ERP-Lösungen erschlagen mich mit Funktionen, die ich nicht brauche, und gibt es nur im Abo und/oder ohne Möglichkeit des Hostings auf dem eigene Proxmox. Ich bevorzuge es, die Software einmal zu kaufen und nur bei wirklich nützlichen Änderungen oder EOL ggf den Nachfolger zu erwerben, monatliches Abo, ohne tatsächlichen Besitz und mit Verlust aller Daten nach Vertragsende, sind nicht so meins.
Kraft scheint für meine Zwecke grundsätzlich brauchbar zu sein, deswegen teste ich es ja gerade. Ich schreibe Rechnungen erst, wenn grob 1000 € zusammengekommen sind, so sprechen wir hier nur von überschaubaren Anzahlen von 100-150 Rechnungen im Jahr.
Soviel zur Vorgeschichte.

Ggf mache ich mich zum Affen, weil ich einfach zu blöd bin, aber eventuell gibt es diese Funktion einfach nicht, daher frage ich mal nach:

Meine Lieferanten haben täglich wechselnde Preise, es gibt zum einen eine EDIS-Schnittstelle, zum Anderen kann ich auch einfach eine csv-Datei exportieren, um diese in die Warenwirtschaft zu importieren.
Bevor ich also vor jedem Angebot / Rechnung (Angebote sind selten, meine Kunden wollen das Problem gelöst haben und nicht erste eine Woche über Preise diskutieren. Zumal man dieser Tage nur noch über Lieferzeiten diskutiert, Preisdiskussionen finden nicht statt) die nötigen Artikel "zu Fuß" eintrage, wäre es doch einfacher, diese automatisiert zu importieren (auch von mehreren Lieferanten zwecks Preisvergleich). 
Ich kann zwar Kataloge anlegen, aber ich habe keine Möglichkeit, CSV o.ä. zu importieren. Was mache ich falsch? Ich kann gerne eine Datei mit einigen Positionen als Beispiel zur Verfügung stellen.

Bitte mich mal in die richtige Richtung schubsen. Danke!

Ebenso, aber nicht so wichtig: wenn ich an zwei Standorten Kraft nutzen will, reicht es, die mysql-Datenbank per VPN erreichen zu können oder muss ich den Kraft-Ordner per Nextcloud synchronisieren?

Gruß,
Martin

**Comment by dragotin** [comment]
Hallo Martin, 

danke, dass du geschrieben hast, hier sollte es kein "zum Affen machen" geben, diese Fragen gehören auf den Tisch und ich gebe gern meinen Senf dazu.

## Kosten
Grundsätzlich ist Kraft unter der GPL als freie Software released. Das heisst, dass jeder den Sourcecode ohne direkte Zahlungen verwenden kann, umsonst sozusagen ;-) Die GPL Lizenz sichert aber keine Kostemfreiheit zu, sondern vor allem einige Freiheiten im Umgang mit dem Sourcecode. Bei Interesse empfehle ich einschlägige Artikel, die zahlreich im Internet rumschwirren.

Üblich ist es mit freier Software, für Services um die Software zu bezahlen: Beispiele wären bestimmte Service-Angebote zur Software wie Einrichtungs-, Update- oder Support-Services, Hosting oder soetwas. Sowas gibt es für Kraft im Moment auch noch nicht, wäre aber zb. für die Erstellung von Costom-Templates als Dienstleistung denkbar.

 ## Katalog Import

Das wurde schon einige Male nachgefragt, und andiskutiert, es ist aber bisher nicht implmenentiert und daher nicht möglich. Dafür war die Nachfrage bislang zu klein. Denn: Die Idee von Kraft ist, wirklich einfach und nützlich zu sein, und nicht in der Menge von Features unübersichtlich zu werden. 

Das von Dir erwähnte EDIFACT ist aber ziemlich kompliziert, so dass das bisher nicht drin ist, weil auch das Katalogmodell von Kraft wesentlich einfacher ist.

Ich kann aber gern mal in das CSV-Format gucken, vielleicht kann man sich einen "good enough" Variante ausdenken. Bitte schicke mir mal eine Test-Datei per link oder email an kraft at freisturz.de 

## Verteilte Kraft

Was dazu schon geht, kann man im Kraft Handbuch nachlesen: https://dragotin.github.io/krafthandbook/manuals/kraft-de.html#_using_kraft_collaboratively

Es ist mein nächstes größeres Ziel für die Entwicklung von Kraft, das noch besser zu machen, zB. mit cloud Integration.

**Comment by noseshimself** [comment]
> Ebenso, aber nicht so wichtig: wenn ich an zwei Standorten Kraft nutzen will, reicht es, die MySQL-Datenbank per VPN erreichen zu können oder muss ich den Kraft-Ordner per Nextcloud synchronisieren?

Vorweg: Da mir für die reiner Office-Anwendung meine Qubes-Umgebung zu sperrig ist, nehme ich dafür ChromeOS, bei dem die Anwender-Linux-Umgebungen in Containern innerhalb einer VM liegen, die abgeschirmt im eigentlichen ChromeOS unterwegs ist. Nextcloud läuft im Container für Office-Anwendungen, synchronisiert aber nur das allernotwendigste, für alles andere habe ich einen Android-Client.

Nextcloud ist schon notwendig und es lohnt sich, einige der Mechanismen von Kraft dabei auszuhebeln.

Ich habe in Nextcloud einen Baum "Kraft" freigegeben, der auf allen Maschinen synchronisiert oder per WebDAV gemounted ist. Darin befinden sich für alle Rechner die Verzeichnis(-bäume)
<Benutzer>/<Maschine>
Archiv
CI
Fonts (ich verwende lokale offline-Fonts für die Dokumenterstellung, weil das massiv schneller ist)
PDF
reports

Unter <Benutzer>/myidentity.vcd hat jeder seine Version der Firmen-Adresse, da ich die Adresse als Lieferadresse auf Ausdrucken angebe und Kraft noch keine Vorkehrungen hat, das pro Dokument benutzerabhängig spezifizieren zu können. Dafür ist auf Briefpapier die Postanschrift "hardcoded"
Unter <Maschine> liegen jeweils die spezifischen kraftrc und kraftdatabaserc für den Arbeitsplatz des Benutzers, mit links in .config/kraftrc und .config/krafdatabasetrc -- damit ist es möglich, den Nextcloud-Baum auf jedem Rechner anders zu nennen und zu positionieren. Environment-Variablen wären schöner gewesen, aber...

```
[document]
ShowDocumentLocale=true

[reporting]
PdfOutputDir=/[PFAD]/Kraft/PDF

[system]
DoXmlArchive=true
XmlArchivePath=/[PFAD]/Kraft/Archiv
mailUA=thunderbird

[userdefaults]
DateFormat=ISO
InsertTemplChapterName=Sonstige
doctype=Rechnung
greeting=
salut=-1
```

In Archiv legt Kraft seine XML-Kopien gedruckter Rechnungen ab. in CI liegen die PDF-Hintergründe meines Briefpapiers, in PDF dir gedruckten Dokumente und in reports die Druckvorlagen. 

Ich bin allen Problemen aus dem Weg gegangen und habe ein slackhq/Nebula Mesh, damit ist das Problem, wo welcher Rechner, Drucker oder sonstwas ist, gar nicht; für die Zwecke verwende ich die entsprechenden Namen.

**Comment by noseshimself** [comment]
> Meine Lieferanten haben täglich wechselnde Preise, es gibt zum einen eine EDIS-Schnittstelle, zum Anderen kann ich auch einfach eine csv-Datei exportieren, um diese in die Warenwirtschaft zu importieren.

Jetzt habe ich tatsächlich in die Datenbankstruktur und in die Quellen gesehen... Hm. An der Datenstruktur würde ich nichts ändern, aber wenn Du Deine Input-Daten in den vorhandenen Feldern unterbringen kannst, könntest Du Dir ein "Kapitel" einrichten und, falls die Daten in irgendeinem brauchbaren Format für die schnelle Weiterverarbeitung (CSV, XNL...) vorlägen mit wenig Programmieraufwand in die Datenbank transportieren. Im einfachsten Fall kannst Du ja vor dem Befüllen alle betreffenden Zeilen löschen und jeden Tag einen neuen Import ablaufen lassen (für Updates müßtest Du für jeden Artikel prüfen, ob er schon da ist -- und da ich ein paar Katalog-erstellende Firmen kenne, habe ich keine großen Erwartungen an die Konsistenz der Schlüssel mehr, da wecheln schon mal die Artikelnummern und -bezeichnungen wild hin-und her). Der Fantasie ist auch keine Grenze gesetzt, wenn Du das tun möchtest, kannst Du auch ein SQL-Script als Zwischenstufe "compilieren" und das per mysql-Client ausführen lassen.

Wenn ich wirklich in dieser Situation wäre, würde ich wahrscheinlich diesen Weg gehen und die Import-Funktion nicht in die Anwenderoberfläche integrieren...

**Reply by e-e-e-tech** [reply]
Hallo,

das Thema ist zwar schon etwas älter, aber ich möchte es mal aufgreifen.
Vielen Dank erst einmal für die freiwillige Arbeit an einem solchen Projekt und die Möglichkeit für die Allgemeinheit, das Programm zu nutzen.
Ich betreibe eine kleine Elektrofirma und verwende fast ausschließlich Open Source Software weil ich einmal mit einem proprietärem Programm heftig auf die Nase gefallen bin und in einer Sackgasse endete. Eine Ausnahme gibt es noch: meine Auftragsbearbeitung.
In unserer Branche gibt es für die Leistungs- bzw. Artikelkataloge langjährig (teilweise über 30 Jahre) stabile Formate (ZVEHnorm, DKEnorm und DATAnorm) für den Import in die ERP-Programme. Hier würde sich wahrscheinlich die Programmierung eines Importfilters lohnen. Meine Programmierkenntnisse sind seit meinem Studium komplett eingestaubt, aber vielleicht hat ja jemand sich schon einmal mit dem Thema näher befasst, würde mich sehr freuen, eine Rückmeldung zu erhalten.

Viele Grüße

**Comment by noseshimself** [comment]
> In unserer Branche gibt es für die Leistungs- bzw. Artikelkataloge langjährig (teilweise über 30 Jahre) stabile Formate (ZVEHnorm, DKEnorm und DATAnorm) für den Import in die ERP-Programme. Hier würde sich wahrscheinlich die Programmierung eines Importfilters lohnen. Meine Programmierkenntnisse sind seit meinem Studium komplett eingestaubt, aber vielleicht hat ja jemand sich schon einmal mit dem Thema näher befasst, würde mich sehr freuen, eine Rückmeldung zu erhalten.

Du hast Dir sicher Datanorm-Datensätze schon angesehen. Klar, das ist so ein CSV-artiges Format, sicherlich nicht wild, die Trennzeichen zu suchen und die Felder zu finden

Und dann?

Die Felder enthalten auch Referenzen auf weitere Daten (Bilder, PDF, ggf. auch mehr). Den ganzen Kram muß man natürlich validieren (PDF -- klar, holen wir uns hinterher online etwas Schadsoftware) und sich überlegen, wo er denn hin soll.

Ok, die Daten sind jetzt in irgendeine Datenbank eingelesen. Und jetzt? Wie sollen die in Kraft zu den vorhandenen (eher wenigen) Datenfelder pro Artikel zugeordnet werden? Das kommt sicher auch auf die Quelle an; unterschiedliche Datenquellen befüllen nicht unbedingt alles ähnlich. Das wird dann schon einiges an Programmieraufwand. Ich bin mir sicher, daß das jemand gegen Einwurf mittelgroßer Münzen zu tun bereit sein wird, aber das dürfte eine massive Erweiterung des Programms werden. Ich kann mich auch irren, ich habe von Anwendungsprogrammierung erfrischend wenig Ahnung.

**Comment by dragotin** [comment]
Die Grundidee von Kraft ist, einfach und nützlich zu sein und nicht zu einem Tool zu werden, das 1000 Möglichkeiten bietet, die aber für 99% der Anwender nicht nützlich sind. 

Das heisst nicht, dass deshalb Funktionen nicht da sein dürfen, aber sie sollten denen nicht im Weg stehen, die sie nicht benutzen wollen.

Auf jeden Fall ist das der Grund, warum es noch keine Importmöglichkeit für DataNorm and Friends gibt, und auch, warum das Katalog-Format in Kraft sehr einfach ist und nur eine Beschreibung und wenig Zusatzdaten umfasst. Es ist eben bisher nur auf "handgepflegte" Kataloge abgestimmt.

Folgende Fragen müssten geklärt werden:

1. Kann man den Inhalt des Datanorm Katalogs sinnvoll auf den Kraft-Katalog abbilden, oder welche zusätzlichen Felder müssten die Kraft-Kataloge haben?
2. Gehts um einmaligen Import der Daten, oder sollen die Daten auch geupdatet werden?
3. Reichen die Struktur-Elemente der Kraft Kataloge (Kapitel und Unterkapitel) aus, um die Menge der Datanorm-Einträge zu verwalten?

Was mich darüberhinaus interessieren würde: 
- Benötigst Du wirklich ALLE Einträge aus dem Datanorm Katalog, oder ist da in der Praxis nur ein Bruchteil der Einträge relevant?
- Sind die Vorlagen, die Du da kriegst, in einer Qualität, dass du sie direkt verwenden kannst und willst?

Ich denke schon, dass man die Vorlagen-Kataloge von Kraft ausbauen könnte, und auch Importe aus solchen Dateien berücksichtigen könnte. Aber wie @noseshimself schon sagt, das ist viel Arbeit und müsste sich dann schon wirklich lohnen.

**Comment by e-e-e-tech** [comment]
Hallo,

vielen Dank für die Antworten.
Bei den Datanorm-Datensätzen wäre ich nur an reinem Text/Zahlen interessiert, ob es da auch Bilderreferenzen o.ä. gibt, weiß ich nicht, wäre mir auch egal.
Die Importfilter kann man ja als Module nachinstallierbar machen, so dass jemand der es simpel mag, die Filter außen vor lassen kann.
Inwieweit eine Abbildung des Data-/DKE-/ZVEHnorm-Formates auf das Kraftformat möglich ist, kann ich leider nicht sagen, oder ob da noch größere Anpassungen nötig wären.
Prinzipiell wären die vorhandenen Strukturelemente in Kraft ausreichend, wenn eine Vollindizierung der Texte und eine Vollindexsuche möglich ist. 
Im Datanorm-Format gibt es den Artikelimport (initial) und die Preisanpassung (mit Artikelanpassung periodisch) bei mir alle 3-6 Monate notwendig.
ZVEH- und DKEnorm bringen jährlich neue Kataloge (man kann aber auch 5 Jahre mit einem Katalog arbeiten, da Arbeitszeiten hinterlegt sind und die Stundensatzkalkulation separat stattfindet, Materialpreise lassen sich händisch anpassen - denke ich - oder prozentual für das jeweilige Dokument).
Natürlich braucht man nicht alle Datensätze der jeweiligen Kataloge, das weiß man aber dummerweise erst im Bedarfsfall, also besser mehr als zu wenig.
Last but not ....., ich bewerfe schon regelmäßig div. Open Source Projekte mit Scheinen und würde das, für den Fall der Importfilter, bei dem der sich die Mühe macht, ebenso handhaben, unter der Voraussetzung, dass das Ergebnis unter die GPL gestellt wird (ich bin für Vorschläge offen).
Schöne Grüße

**Reply by noseshimself** [reply]
> ZVEH- und DKEnorm bringen jährlich neue Kataloge

Manche Großhändler bringen heute zweimal pro Woche Preisupdates (und die Kataloge können auch mal 100000 Datensätze enthalten)... Da ist das dann schon eine andere Größenordnung.

**Comment by dragotin** [comment]
Hallo,

danke für die Diskussion. Es ist wirklich nicht leicht, weil das Fehlen solcher Features sicher auch viele potentielle Benutzer abschreckt.

Andererseits, wie auch schon weiter oben gesagt: Irgendwie passt es für mich nicht zur Grundidee von Kraft. Einfach Unmengen von Daten in Kraft reinzuimportieren, nur falls man sie irgendwann mal braucht, ist doch nicht effizient?! Ist es da nicht einfacher, die Datensätze, die wirklich gebraucht werden, einmalig in Kraft zu übertragen und einmal im Jahr Preise anzupassen?

Die Struktur der Kraft-Kataloge müsste für so einen (vollumfänglichen) Import deutlich verändert werden. Das kann man machen, solang die betont einfache Verwendung wie sie jetzt ist dadurch nicht erschwert wird. So einen Import würde ich dann durch ein externes tool machen, der Kraft-kompatible Kataloge erstellt, die dann in Kraft eingebunden werden.

Das ist etwas Arbeit, die aber nicht so sehr ausserhalb der Roadmap liegt. Ich würde es eh gut finden, wenn man Kataloge hinzufügen könnte, und ggf über ein XML Format zb. weitergeben könnte.  

Bezgl. Bezahlung: Es ist ein sehr guter Ansatz, solche Features durch bezahlte Entwicklung erstellen zu lassen. Als Maintainer von Kraft akzeptiere ich das auch nur unter GPL Lizenz. Die Frage ist, ob man genug Budget zusammentragen kann, um zb. das Datanorm Feature zu finanzieren. Zur Koordination und fürs technische Konzept stehe ich natürlich gern zur Verfügung wenn das jemand anpacken möchte.

Noch als Referenz: Hier ist ein ganz interessantes Tool um DATANORM4 nach CSV zu exportieren: https://github.com/yNiklas/datanorm/tree/main
